sky/utils/common_utils.py

```diff
@@ -792,4 +789,69 @@ def get_mem_size_gb() -> float:
             with ux_utils.print_exception_no_traceback():
                 raise ValueError(
                     f'Failed to parse the memory size from {mem_size}') from e
-    return psutil.virtual_memory().total / (1024**3)
+    # host memory limit
+    mem = psutil.virtual_memory().total
+    cgroup_mem = _get_cgroup_memory_limit()
+    if cgroup_mem is not None:
+        mem = min(mem, cgroup_mem)
+    return mem / (1024**3)
+
+
+# Refer to:
+# - https://docs.kernel.org/admin-guide/cgroup-v1/index.html
+# - https://docs.kernel.org/admin-guide/cgroup-v2.html
+# for the standards of handler files in cgroupv1 and v2.
+# Since all those paths are well-known standards that are unlikely to change,
+# we use string literals instead of defining extra constants.
+def _get_cgroup_cpu_limit() -> Optional[float]:
+    """Return cpu limit from cgroups in cores.
+
+    Returns:
+        The cpu limit in cores as a float (can be fractional), or None if not
+        found, unspecified, or invalid.
+    """
+    try:
+        if _is_cgroup_v2():
+            with open('/sys/fs/cgroup/cpu.max', 'r', encoding='utf-8') as f:
+                quota_str, period_str = f.read().strip().split()
+                if quota_str == 'max':
+                    return None
+                quota = float(quota_str)
+                period = float(period_str)
+                return quota / period if quota > 0 else None
+        else:
+            # cgroup v1
+            quota_path = '/sys/fs/cgroup/cpu/cpu.cfs_quota_us'
+            period_path = '/sys/fs/cgroup/cpu/cpu.cfs_period_us'
+
+            with open(quota_path, 'r', encoding='utf-8') as f:
+                quota = float(f.read().strip())
+            with open(period_path, 'r', encoding='utf-8') as f:
+                period = float(f.read().strip())
+            return quota / period if quota > 0 else None
+    except (IOError, ValueError):
```

Comment:
Would there be `FileNotFoundError` exception?

Reply:
IOError is an alias of OSError, which covers `FileNotFoundError`